qiskit/pulse/ir/ir.py

```diff
@@ -0,0 +1,375 @@
+
+        """
+        for element in self._elements:
+            if isinstance(element, IrBlock):
+                return True
+        return False
+
+    def add_element(
+        self,
+        element: IrElement | List[IrElement],
+    ):
+        """Adds IR element or list thereof to the ``IrBlock`` object.
+
+        Args:
+            element: `IrElement` object, or list thereof to add.
+        """
+        if not isinstance(element, list):
+            element = [element]
+
+        self._elements.extend(element)
+
+    def shift_initial_time(self, value: int):
+        """Shifts ``initial_time`` of the ``IrBlock`` elements
+
+        The ``initial_time`` of all elements in the ``IrBlock`` are shifted by ``value``,
+        including recursively if any element is ``IrBlock``.
+
+        Args:
+            value: The integer value by which ``initial_time`` is to be shifted.
+
+        Raises:
+            PulseError: if the object is not scheduled (initial_time is None)
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not shift initial_time of IrBlock with unscheduled elements")
+
+        for element in self.elements:
+            element.shift_initial_time(value)
+
+    def flatten(self) -> None:
+        """Flatten the ``IrBlock`` into a single block (in place)
+
+        Recursively flattens all child ``IrBlock`` until only instructions remain.
+
+        Raises:
+            PulseError: If the block is not scheduled.
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not flatten an unscheduled IrBlock")
+
+        new_elements = self._get_flatten_elements()
+        self._elements = new_elements
+
+    def _get_flatten_elements(self) -> List[IrInstruction]:
+        """Return a list of flatten instructions
+
+        Recursively lists all instructions in the ``IrBlock``.
+        """
+        flatten_elements = []
+        for element in self._elements:
+            if isinstance(element, IrBlock):
+                flatten_elements.extend(element._get_flatten_elements())
+            else:
+                flatten_elements.append(element)
+        return flatten_elements
+
+    def remove_instruction(self, instruction: IrInstruction) -> None:
+        """Remove instruction from ``IrBlock`` (in place)
+
+        If the instruction does not exist in the object, no change is made and no error is raised.
+        Because the order of elements in the block is instrumental to scheduling, instructions can't
+        be removed if the block is not scheduled.
+
+        Args:
+            instruction: The instruction to be removed.
+
+        Raises:
+            PulseError: If the ``IrBlock`` is not scheduled.
+        """
+        if self.initial_time is None:
+            raise PulseError("Can not remove instruction from unscheduled IrBlock")
+        if instruction in self._elements:
+            self._elements.remove(instruction)
```

Comment:
I hesitate to update the data by IR itself. Also this implementation is problematic because Python list remove only removes only one value when multiple values exist. You don't consider updating the instruction initial time, here user may expect to shift the rest of instructions after removal (and other users may not). Since the IR is created _after_ the user built the pulse program, I don't think remove method is necessary.

Reply:
I agree there are some problems here. I tried to avoid some of them by limiting this to scheduled IrBlocks. I'll remove it for the time being, but eventually I think we'll need some removal mechanism (method\pass), for example to remove instructions which were broadcasted.